src/HttpClient/Plugin/Authentication/HybridOauth2.php

```diff
@@ -0,0 +1,129 @@
+<?php
+
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Apigee\Edge\HttpClient\Plugin\Authentication;
+
+use Apigee\Edge\Client;
+use Apigee\Edge\ClientInterface;
+use Apigee\Edge\Exception\HybridOauth2AuthenticationException;
+use Firebase\JWT\JWT;
+use Http\Client\Exception;
+
+/**
+ * HybridOauth2 authentication plugin for authenticating to Hybrid Cloud API.
+ *
+ * @see https://developers.google.com/identity/protocols/OAuth2ServiceAccount
+ */
+class HybridOauth2 extends AbstractOauth
+{
+    /**
+     * Authorization server for Apigee Hybrid authentication.
+     *
+     * @var string
+     */
+    public const DEFAULT_AUTHORIZATION_SERVER = 'https://oauth2.googleapis.com/token';
+
+    /**
+     * A space-delimited list of the permissions that the application requests.
+     *
+     * @var string
+     */
+    public const TOKEN_SCOPES = 'https://www.googleapis.com/auth/cloud-platform';
+
+    /**
+     * Grant type used in an access token request.
+     *
+     * @see https://developers.google.com/identity/protocols/OAuth2ServiceAccount
+     *
+     * @var string
+     */
+    public const GRANT_TYPE = 'urn:ietf:params:oauth:grant-type:jwt-bearer';
+
+    /**
+     * The service account email.
+     *
+     * @var string
+     */
+    protected $email;
+
+    /**
+     * The service account private key.
+     *
+     * @var string
+     */
+    protected $privateKey;
+
+    /**
+     * Hybrid Oauth2 constructor.
+     *
+     * @param string $email
+     *   The service account email.
+     * @param string $privateKey
+     *   The service account private key.
+     * @param \Apigee\Edge\HttpClient\Plugin\Authentication\OauthTokenStorageInterface $token_storage
+     *   Storage where access token gets saved.
+     * @param string|null $authServer
+     *   Authentication server.
+     */
+    public function __construct(string $email, string $privateKey, OauthTokenStorageInterface $token_storage, ?string $authServer = null)
+    {
+        $this->email = $email;
+        $this->privateKey = $privateKey;
+        parent::__construct($token_storage, $authServer ?: self::DEFAULT_AUTHORIZATION_SERVER);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    protected function authClient(): ClientInterface
+    {
+        return new Client(new NullAuthentication(), $this->authServer);
```

Comment:
Well, there are two possible solutions here:
1. What you did, introducing a new NullAthentication plugin.
2. Make the first parameter optional. As I can see, the client's implementation can already handle that. `\Apigee\Edge\Client::getDefaultPlugins()`.

If we stick the current solution, `\Apigee\Edge\Tests\Test\HttpClient\Plugin\NullAuthentication` can be removed (there is no BC for test files) because it is a duplicate. If we would go with the 2nd option then we would remove an important constraint for non-hibrid orgs. On those orgs you must provide an authentication plugin (usually basic auth) to be able to call an endpoint, let it be an API endpoint or the OAuth authorization server.)

Reply:
Let's keep (1), as it keeps the constraint for the non-hybrid orgs. I've removed the `NullAuthentication` class from the tests and replaced it with this one.